packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -37,59 +39,130 @@ impl HookContextImpl {
 
     pub(crate) fn as_hook_context<T>(&mut self) -> HookContext<'_, T> {
         HookContext {
-            parent: self,
+            inner: HookContextInner::new(self),
             _marker: PhantomData::default(),
         }
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct HookContextInner<'a> {
+    inner: &'a mut HookContextImpl,
+
+    emits: Vec<Emit>,
+    snippets: Vec<Emit>,
+}
```

Comment:
Why we need another struct here? Also, when `HookContextImpl` needs to be a. `&mut`? Removing the reference would remove the lifetime on `HookContext` as well and I don't see why we ever need to use `HookContextImpl` at all, this should be an implementation detail to allow casting with`repr(transparent)`.

Reply:
This is because there are 2 different layers here at play: Debug level invocation and Hook level invocation.

`HookContextInner` is instantiated for every hook, but `HookContextImpl` is only instantiated once for every `Debug` invocation.

This is why `HookContextImpl` hosts `alternate` and `storage`, while `HookContextInner` only has emits and snippets.

The `snippets` property is here to make `return` statements a bit easier in the debug render code, as otherwise, the conditional compilation step would be an utterly unreadable mess, basically we destruct `HookContextInner`, reorder the snippets and emits, then append the snippets to the `HookContextImpl::snippets` and return the emits.

`storage` is the same across the whole `Debug` invocation, not only hook, you can add new data to it, which means it needs to be `mut`.

We could, in theory, move this into `HookContext`, but then wouldn't be able to `repr(transparent)` cast.

Reply:
I see. I experimented around with this a little bit and moved everything from `HookContextImpl` to `HookContextInner` (so removing the lifetime). I changed the functions to take `HookContext<Frame>` instead of `HookContextImpl`. Instead of `as_hook_context` I reused the `HookContext`, but changed the `into_parts` method to
```rust
    fn take_emits(&mut self) -> (Vec<Emit>, Vec<Emit>) {
        (mem::take(&mut self.emits), mem::take(&mut self.snippets))
    }
```

I opened a PR for this change, so you can see my proposed changes 🙂 
- https://github.com/indietyp/hash/pull/2

Reply:
PR was merged 🎉